x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java

```diff
@@ -285,37 +338,174 @@ void persistStateToClusterState(DataFrameTransformState state,
     @Override
     public synchronized void onCancelled() {
         logger.info(
-                "Received cancellation request for data frame transform [" + transform.getId() + "], state: [" + indexer.getState() + "]");
-        if (indexer.abort()) {
+                "Received cancellation request for data frame transform [" + transform.getId() + "], state: [" + taskState.get() + "]");
+        if (getIndexer() != null && getIndexer().abort()) {
             // there is no background transform running, we can shutdown safely
             shutdown();
         }
     }
 
-    protected class ClientDataFrameIndexer extends DataFrameIndexer {
-        private static final int LOAD_TRANSFORM_TIMEOUT_IN_SECONDS = 30;
+    synchronized void initializeIndexer(ClientDataFrameIndexerBuilder indexerBuilder) {
+        indexer.set(indexerBuilder.build(this));
+    }
+
+    static class ClientDataFrameIndexerBuilder {
```

Comment:
This class was necessary as the executor gathers the necessary dependencies in individual async calls. I needed a place to store them. Initially I kept them in local state in the executor, but that conflated ownership.